src/v/cloud_storage/tests/segment_meta_cstore_test.cc

```diff
@@ -1342,3 +1344,551 @@ BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_append_retrieve_edge_case) {
       = metadata[cloud_storage::cstore_max_frame_size * 2].base_offset;
     BOOST_CHECK_NO_THROW(store.lower_bound(last_frame_first_offset));
 }
+
+class segment_meta_random_walk {
+    static constexpr int64_t ts_start = 1700000000000;
+
+public:
+    segment_meta_random_walk(
+      bool add_reuploads,
+      bool add_reorderings,
+      bool add_gaps,
+      bool add_overlaps,
+      bool add_empty)
+      : _add_reuploads(add_reuploads)
+      , _add_reorderings(add_reorderings)
+      , _add_gaps(add_gaps)
+      , _add_overlaps(add_overlaps)
+      , _add_empty(add_empty) {}
+
+    /// Returns truncate-offset, expected-start-offset
+    /// Offsets might be different in case of out of order
+    /// truncations.
+    std::pair<model::offset, model::offset> maybe_truncate() {
+        if (
+          !_ooo_write.empty() ||
```

Comment:
I'm wondering what this translates to when thinking about operations to the `archival_metadata_stm`. If these writes haven't made their way to the cstore, what's the harm in truncating?